lib/Sema/TypeCheckDecl.cpp

```diff
@@ -913,18 +914,101 @@ IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
+  // The legacy check
+  if (!ctx.LangOpts.hasFeature(Feature::NoncopyableGenerics)) {
+    if (isa<ClassDecl>(decl) || isa<StructDecl>(decl) || isa<EnumDecl>(decl)) {
       if (decl->getAttrs().hasAttribute<MoveOnlyAttr>()) {
         if (!decl->getASTContext().supportsMoveOnlyTypes())
-            decl->diagnose(diag::moveOnly_requires_lexical_lifetimes);
+          decl->diagnose(diag::moveOnly_requires_lexical_lifetimes);
 
         return true;
       }
+    }
+    return false;
   }
-  return false;
+  // ----------------
+
+  // FIXME: just never allow lexical-lifetimes to be disabled?
+  if (!ctx.supportsMoveOnlyTypes())
+    decl->diagnose(diag::moveOnly_requires_lexical_lifetimes);
+
+  if (isa<TypeAliasDecl>(decl)) {
+    llvm_unreachable("todo: handle type aliases");
+  }
+
+  assert(isa<NominalTypeDecl>(decl) && "todo: handle non-nominals");
+
+  // For protocols, inspect its generic signature for a Copyable requirement
+  if (auto *proto = dyn_cast<ProtocolDecl>(decl)) {
+    auto *copyableProto = ctx.getProtocol(KnownProtocolKind::Copyable);
+    if (!copyableProto)
+      llvm_unreachable("missing Copyable protocol!");
+
+    auto sig = proto->getGenericSignature();
+    auto requiresCopyable =
+        sig->requiresProtocol(proto->getSelfInterfaceType(), copyableProto);
+
+    // TODO: report that something implied Copyable despite writing ~Copyable?
+
+    return !requiresCopyable;
+  }
+
+  // Handle the legacy '@_moveOnly' attribute
+  if (decl->getAttrs().hasAttribute<MoveOnlyAttr>()) {
+    assert(isa<StructDecl>(decl) || isa<EnumDecl>(decl));
+    return true;
+  }
+
+  // For all other nominals, we can simply check the inheritance clause.
+
+  // The set of defaults all types start with.
+  auto defaults = getInvertibleProtocols();
+  std::map<KnownProtocolKind, SourceLoc> lastInverse;
+
+  // Does correctness checking after inspecting the inheritance clause.
+  // TODO: we can probably move this to `checkInheritanceClause` instead.
+  auto foundInverse = [&](SourceLoc loc, KnownProtocolKind kind) {
+    // do we still have this default?
+    if (defaults.contains(kind)) {
+      defaults.remove(kind);
+      lastInverse.insert({kind, loc});
+
+      // Classes are always copyable.
+      if (isa<ClassDecl>(decl) && kind == KnownProtocolKind::Copyable) {
+        ctx.Diags.diagnose(loc, diag::noncopyable_class);
+        return;
+      }
+
+      return;
+    }
+
+    // diagnose duplicate inverses
+    ctx.Diags.diagnose(loc, diag::inverse_duplicate);
+    ctx.Diags.diagnose(lastInverse[kind], diag::inverse_duplicate_previous);
+  };
+
+  // Check the inheritance clause for inverses.
+  auto inherited = decl->getInherited();
+  for (size_t i = 0; i < inherited.size(); i++) {
+    auto *repr = inherited.getTypeRepr(i);
+
+    if (!dyn_cast_or_null<InverseTypeRepr>(repr)) {
+      continue;
+    }
+
+    auto type = inherited.getResolvedType(i, TypeResolutionStage::Structural);
+    auto kp = type->getKnownProtocol();
```

Comment:
Is it really worth adding a new TypeBase::getKnownProtocol()? It appears to only used in this one place.

Reply:
I found 3 and then added another use to help clean up some code. I kind of like it. 🤷🏼